src/csharp/Grpc.Core/Channel.cs

```diff
@@ -252,6 +260,7 @@ public async Task ShutdownAsync()
             lock (myLock)
             {
                 handle.Dispose();
+                cleanUpCredentials?.Invoke();
```

Comment:
this is not the right time to cleanup the native callback.

verify_peer_destruct  from verify_peer_options should provide the signal to free the gchandle.
refer to https://github.com/grpc/grpc/blob/b3b5d634231ce2c5c0ec0c557b6844e1a43b482e/src/csharp/Grpc.Core/Internal/NativeMetadataCredentialsPlugin.cs#L57  for an example.
IMHO there should be no changes in Channel.cs at all necessary.

Reply:
Indeed using `verify_peer_destruct` is much better, hadn't thought of that. Updated to use it.

Reply:
Actually realized that I also needed a gcHandle for the destruct callback. But looking at it, I must say I am having trouble to understand how things work, even for the example your provided.

In `NativeMetadataCredentialsPlugin`, the `NativeMetadataInterceptorHandler` references `gcHandle` which is a field in the class. But if an instance of the class gets GC'd, wouldn't it be collected as well? Or does it mean that the entire object will not be collected until the `GCHandle` is realeased?

I tried to look for documentation but could not find much, I just want to make sure that my implementation makes sense.

Reply:
In NativeMetadataCredentialsPlugin I'm actually squashing more native callbacks into just one with more parameters. 

Ad releasing:   gcHandle is on the delegate object, but the delegate implicitly reference and instance of the class ("this") so the class can't get garbage collected as long as the gcHandle is allocated. (I think referencing the class instance itself in the GCHandle would work too).

Reply:
I hadn't looked at the native part for your example, sorry for that, I think I got it now.

Thanks for the answer regarding `gcHandle`. However I'm still a bit curious, doing some reading on the subject, I found [this article ](https://www.microsoftpressstore.com/articles/article.aspx?p=2224054&seqNum=4) that seems to say that you're supposed to pass the `gcHandle` to the native code and not the object you want to callback on in case of the object being moved by memory compaction during GC:
> You can’t actually pass a pointer to a managed object out to native code, because if a garbage collection occurs, the object could move in memory, invalidating the pointer. So to work around this, you would call GCHandle’s Alloc method, passing in a reference to the object and the Normal flag. Then you’d cast the returned GCHandle instance to an IntPtr and pass the IntPtr into the native code.

Is that not an issue in this case?

Reply:
I haven't read it in full, but it seems to talk about a different case (when you're passing a pointer to managed object to the native code and the native code directly accesses the representation of that object in memory). We're not doing that, we're only passing a delegate and the marshalling semantics of P/Invoke in that case is to translate the delegate into a native callback and pass that one to the native code (this transitions object is called the "thunk" in the docs btw).

Reply:
Thanks for the thorough answer!